### PR TITLE
Add TF for vpc endpoint on GCP.

### PR DESCRIPTION
**Pull request by jessiedu-db**
Add TF for vpc endpoint on GCP.

### CODE REVIEW

File: mws/mws.go [hunk 1]
```diff
@@ -72,17 +72,27 @@ type Network struct {
...
+	VPCEndpointID           string              `json:"vpc_endpoint_id,omitempty" tf:"computed"`
+	AwsVPCEndpointID        string              `json:"aws_vpc_endpoint_id"`
+	AccountID               string              `json:"account_id,omitempty"`
+	VPCEndpointName         string              `json:"vpc_endpoint_name"`
+	AwsVPCEndpointServiceID string              `json:"aws_endpoint_service_id,omitempty" tf:"computed"`
+	AWSAccountID            string              `json:"aws_account_id,omitempty" tf:"computed"`
+	UseCase                 string              `json:"use_case,omitempty" tf:"computed"`
+	Region                  string              `json:"region"`
+	State                   string              `json:"state,omitempty" tf:"computed"`
+        GcpVpcEndpointInfo      *GcpVpcEndpointInfo `json:"gcp_vpc_endpoint_info, omitempty"`
```
Comment (nkvuong) [inline mws/mws.go:95]:
```suggestion
        GcpVpcEndpointInfo      *GcpVpcEndpointInfo `json:"gcp_vpc_endpoint_info,omitempty"` 
```

File: docs/resources/mws_vpc_endpoint.md [hunk 2]
```diff
@@ -111,19 +146,26 @@ resource "databricks_mws_workspaces" "this" {
 
 The following arguments are required:
 
-* `account_id` - Account Id that could be found in the bottom left corner of [Accounts Console](https://accounts.cloud.databricks.com/)
-* `aws_vpc_endpoint_id` - ID of configured [aws_vpc_endpoint](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/vpc_endpoint)
+* `account_id` - Account Id that could be found in the [Accounts Console](https://accounts.cloud.databricks.com/)
```
Comment (nkvuong) [inline docs/resources/mws_vpc_endpoint.md:149]:
```suggestion
* `account_id` - Account Id that could be found in the Accounts Console for [AWS](https://accounts.cloud.databricks.com/) or [GCP](https://accounts.gcp.databricks.com/)
```

File: mws/mws.go [hunk 3]
```diff
@@ -72,17 +72,27 @@ type Network struct {
...
-	AwsVPCEndpointID        string `json:"aws_vpc_endpoint_id"`
-	AccountID               string `json:"account_id,omitempty"`
-	VPCEndpointName         string `json:"vpc_endpoint_name"`
-	AwsVPCEndpointServiceID string `json:"aws_endpoint_service_id,omitempty" tf:"computed"`
-	AWSAccountID            string `json:"aws_account_id,omitempty" tf:"computed"`
-	UseCase                 string `json:"use_case,omitempty" tf:"computed"`
-	Region                  string `json:"region"`
-	State                   string `json:"state,omitempty" tf:"computed"`
+	VPCEndpointID           string              `json:"vpc_endpoint_id,omitempty" tf:"computed"`
+	AwsVPCEndpointID        string              `json:"aws_vpc_endpoint_id"`
```
Comment (nkvuong) [inline mws/mws.go:87]:
```suggestion
	AwsVPCEndpointID        string              `json:"aws_vpc_endpoint_id,omitempty"`
```

File: mws/mws.go [hunk 4]
```diff
@@ -72,17 +72,27 @@ type Network struct {
...
+	AccountID               string              `json:"account_id,omitempty"`
+	VPCEndpointName         string              `json:"vpc_endpoint_name"`
+	AwsVPCEndpointServiceID string              `json:"aws_endpoint_service_id,omitempty" tf:"computed"`
+	AWSAccountID            string              `json:"aws_account_id,omitempty" tf:"computed"`
+	UseCase                 string              `json:"use_case,omitempty" tf:"computed"`
+	Region                  string              `json:"region"`
```
Comment (nkvuong) [inline mws/mws.go:93]: this is now optional

```suggestion
	Region                  string              `json:"region,omitempty"`
```

File: mws/resource_mws_vpc_endpoint_test.go [hunk 5]
```diff
@@ -54,6 +54,58 @@ func TestResourceVPCEndpointCreate(t *testing.T) {
...
+		gcp_vpc_endpoint_info {
+			project_id = "project_a"
+			psc_endpoint_name = "endpoint_a"
+			endpoint_region = "region_a"
+        }
+		`,
+		Create: true,
+	}.Apply(t)
+	assert.NoError(t, err)
+	assert.Equal(t, "abc/nid", d.Id())
```
Comment (nkvuong) [inline mws/resource_mws_vpc_endpoint_test.go:106]: Should this assertion be `"abc/ve_id"`

**Review by nkvuong — CHANGES_REQUESTED** [review]

**Comment by nkvuong** [comment]
we will need to add acceptance test for this

File: mws/resource_mws_vpc_endpoint.go [hunk 6]
```diff
@@ -76,6 +76,13 @@ func ResourceMwsVpcEndpoint() *schema.Resource {
 	s := common.StructToSchema(VPCEndpoint{}, func(s map[string]*schema.Schema) map[string]*schema.Schema {
 		// nolint
 		s["vpc_endpoint_name"].ValidateFunc = validation.StringLenBetween(4, 256)
```
Comment (jessiedu-db) [inline mws/resource_mws_vpc_endpoint.go:81]: For line 78 s["vpc_endpoint_name"].ValidateFunc = validation.StringLenBetween(4, 256):
1. AWS and GCP have different restriction on the string length.
2. We have validation check for the input string length for AWS and GCP in the backend. A user error will be reported by the server.

In this case, do we need to add specific override for GCP here? @nkvuong

File: docs/resources/mws_vpc_endpoint.md (hunk 2 above)
Comment (jessiedu-db) [reply docs/resources/mws_vpc_endpoint.md]: DOne.

File: mws/mws.go (hunk 4 above)
Comment (jessiedu-db) [reply mws/mws.go]: Done.

File: mws/resource_mws_vpc_endpoint_test.go (hunk 5 above)
Comment (jessiedu-db) [reply mws/resource_mws_vpc_endpoint_test.go]: Done.

File: mws/mws.go (hunk 3 above)
Comment (jessiedu-db) [reply mws/mws.go]: Done.

File: docs/resources/mws_vpc_endpoint.md [hunk 7]
```diff
@@ -7,12 +7,14 @@ subcategory: "Deployment"
 
 -> **Note** This resource has an evolving API, which will change in the upcoming versions of the provider in order to simplify user experience.
 
-Enables you to register [aws_vpc_endpoint](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/vpc_endpoint) resources with Databricks such that they can be used as part of a [databricks_mws_networks](mws_networks.md) configuration.
+Enables you to register [aws_vpc_endpoint](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/vpc_endpoint) resources or gcp vpc_endpoint resources with Databricks such that they can be used as part of a [databricks_mws_networks](mws_networks.md) configuration.
```
Comment (jessiedu-db) [inline docs/resources/mws_vpc_endpoint.md:10]: @nkvuong @nfx  Can I assume there will be a (https://registry.terraform.io/providers/hashicorp/gcp/latest/docs/resources/vpc_endpoint) after this PS is merged?

File: mws/resource_mws_vpc_endpoint.go (hunk 6 above)
Comment (nkvuong) [reply mws/resource_mws_vpc_endpoint.go:81]: Terraform does not support more complex`ValidateFunc`, so for now we should just set the string length validation to cover both AWS & GCP

File: docs/resources/mws_vpc_endpoint.md (hunk 7 above)
Comment (nkvuong) [reply docs/resources/mws_vpc_endpoint.md:10]: GCP Terraform doc is missing this, so we have to point to this [open issue](https://github.com/hashicorp/terraform-provider-google/issues/11225) for now

**Comment by nkvuong** [comment]
I don't think the changes to `go.mod` and `go.sum` are intended

File: internal/acceptance/mws_vpc_endpoint_test.go [hunk 8]
```diff
@@ -17,3 +17,21 @@ func TestMwsAccVpcEndpoint(t *testing.T) {
 		}`,
 	})
 }
+
+func TestMwsAccVpcEndpoint_GCP(t *testing.T) {
```
Comment (nkvuong) [inline internal/acceptance/mws_vpc_endpoint_test.go:21]:
```suggestion
func TestMwsAccGcpVpcEndpoint(t *testing.T) {
```

File: docs/resources/mws_vpc_endpoint.md [hunk 9]
```diff
@@ -68,25 +70,51 @@ resource "aws_vpc_endpoint" "kinesis-streams" {
...
+Once you have created the necessary PSC connections, you need to register each of them via *this* Terraform resource, which calls out to the Databricks Account API.
 
-resource "databricks_mws_vpc_endpoint" "relay" {
+```hcl
+variable "databricks_account_id" {
+  description = "Account Id that could be found in https://accounts.gcp.databricks.com/"
+}
+variable "databricks_google_service_account" {}
+variable "google_project" {}
+variable "subnet_region"
```
Comment (nkvuong) [inline docs/resources/mws_vpc_endpoint.md:102]:
```suggestion
variable "subnet_region" {}
```

File: internal/acceptance/mws_vpc_endpoint_test.go [hunk 10]
```diff
@@ -17,3 +17,21 @@ func TestMwsAccVpcEndpoint(t *testing.T) {
...
+	accountLevel(t, step{
+		Template: `
+		resource "databricks_mws_vpc_endpoint" "this" {
+			account_id = "{env.DATABRICKS_ACCOUNT_ID}"
+			vpc_endpoint_name = "{env.TEST_PREFIX}-{var.RANDOM}"
+            
+            gcp_vpc_endpoint_info {
+			  project_id = "env.GOOGLE_PROJECT"
+			  psc_endpoint_name = "{var.RANDOM}"
+			  endpoint_region = "env.GOOGLE_REGION"
```
Comment (nfx) [inline internal/acceptance/mws_vpc_endpoint_test.go:32]:
```suggestion
			  endpoint_region = "{env.GOOGLE_REGION}"
```

File: internal/acceptance/mws_vpc_endpoint_test.go [hunk 11]
```diff
@@ -17,3 +17,21 @@ func TestMwsAccVpcEndpoint(t *testing.T) {
...
+func TestMwsAccGcpVpcEndpoint(t *testing.T) {
+	t.SkipNow()
+	accountLevel(t, step{
+		Template: `
+		resource "databricks_mws_vpc_endpoint" "this" {
+			account_id = "{env.DATABRICKS_ACCOUNT_ID}"
+			vpc_endpoint_name = "{env.TEST_PREFIX}-{var.RANDOM}"
+            
+            gcp_vpc_endpoint_info {
+			  project_id = "env.GOOGLE_PROJECT"
```
Comment (nfx) [inline internal/acceptance/mws_vpc_endpoint_test.go:30]:
```suggestion
			  project_id = "{env.GOOGLE_PROJECT}"
```

File: internal/acceptance/mws_vpc_endpoint_test.go [hunk 12]
```diff
@@ -17,3 +17,21 @@ func TestMwsAccVpcEndpoint(t *testing.T) {
 		}`,
 	})
 }
+
+func TestMwsAccGcpVpcEndpoint(t *testing.T) {
+	t.SkipNow()
```
Comment (nfx) [inline internal/acceptance/mws_vpc_endpoint_test.go:22]:
```suggestion
```

**Review by nfx — CHANGES_REQUESTED** [review]
please format files, like `mws/mws.go`

**Review by grayluck — CHANGES_REQUESTED** [review]

File: docs/resources/mws_vpc_endpoint.md [hunk 13]
```diff
@@ -68,25 +70,51 @@ resource "aws_vpc_endpoint" "kinesis-streams" {
 }
 ```
 
-Once you have created the necessary endpoints, you need to register each of them via *this* Terraform resource, which calls out to the [Databricks Account API](https://docs.databricks.com/administration-guide/cloud-configurations/aws/privatelink.html#step-3-register-your-vpc-endpoint-ids-with-the-account-api)):
+Typically the next steps after this would be to create a [databricks_mws_private_access_settings](mws_private_access_settings.md) and [databricks_mws_networks](mws_networks.md) configuration, before passing the `databricks_mws_private_access_settings.pas.private_access_settings_id` and `databricks_mws_networks.this.network_id` into a [databricks_mws_workspaces](mws_workspaces.md) resource:
 
 ```hcl
-resource "databricks_mws_vpc_endpoint" "workspace" {
```
[the remark this answers is not in the document]
Comment (jessiedu-db) [reply docs/resources/mws_vpc_endpoint.md]: This came from a bad version conflict resolution. Thanks!

**Review by nfx — APPROVED** [review]

**Pull request merged**